<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1

In [4]:
odd_squares = []
for number in range(0,10):
    if number %2 == 0: 
        continue
    odd_squares.append(number**2)

print(odd_squares)

[1, 9, 25, 49, 81]


# Exercise 2

In [16]:
[numberz**2 for numberz in range (10) if numberz %2 == 1 ]

[1, 9, 25, 49, 81]

# Exercise 3

In [1]:
%%timeit
odd_squares = []
for number in range(0,10):
    if number %2 == 0: 
        continue
    odd_squares.append(number**2)

1.98 μs ± 287 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [2]:
#note = %%timeit times the whole cell, so even the 
#print() function will affect it immensely 

In [6]:
%%timeit
[numberz**2 for numberz in range (10) if number %2 == 1]

2.34 μs ± 460 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


The one using loops is faster! (edited)

# Exercise 4

In [7]:
radioactive = 100

while True: 
    print (radioactive)
    radioactive *= 0.95
    if radioactive <= 50: break 


100
95.0
90.25
85.7375
81.45062499999999
77.37809374999999
73.50918906249998
69.83372960937498
66.34204312890623
63.02494097246091
59.87369392383786
56.880009227645964
54.03600876626366
51.33420832795048


13 years to reach 51.3, 14 years to reach less than half its original amount 

# Exercise 5

In [7]:
import numpy as np

In [8]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]


In [9]:
for year, CO2_level in co2_data: 
    if (CO2_level) > 3: 
        print(f'{year}, {year - CO2_level}')

2000, 1630.45
2001, 1629.8600000000001
2002, 1628.72
2003, 1627.2
2004, 1626.48
2005, 1625.2
2006, 1624.1
2007, 1623.21
2008, 1622.4
2009, 1621.57
2010, 1620.1
2011, 1619.35
2012, 1618.15
2013, 1616.48
2014, 1615.35
2015, 1614.17
2016, 1611.76
2017, 1610.45
2018, 1609.48
2019, 1607.56
2020, 1605.76


In [10]:
list(enumerate(co2_data))

[(0, (2000, 369.55)),
 (1, (2001, 371.14)),
 (2, (2002, 373.28)),
 (3, (2003, 375.8)),
 (4, (2004, 377.52)),
 (5, (2005, 379.8)),
 (6, (2006, 381.9)),
 (7, (2007, 383.79)),
 (8, (2008, 385.6)),
 (9, (2009, 387.43)),
 (10, (2010, 389.9)),
 (11, (2011, 391.65)),
 (12, (2012, 393.85)),
 (13, (2013, 396.52)),
 (14, (2014, 398.65)),
 (15, (2015, 400.83)),
 (16, (2016, 404.24)),
 (17, (2017, 406.55)),
 (18, (2018, 408.52)),
 (19, (2019, 411.44)),
 (20, (2020, 414.24))]

In [ ]:
for index, data in enumerate(co2_data): #consider using 
    # for index in range(len(co2_data)-1)
    year, CO2_level = data
    next_year, next_co2_level = co2_data[index+1]
    print(next_year,next_co2_level)
    # print(year, CO2_level)
    #print(co2_data[index+1]) # to get the next element 
    # if index == len(co2_data): 
    #     break
    # if n[1] - data[1] >= 3: 
    #     print(data[0])
    break
    
    
    



2001 371.14


In [14]:
for index, data in enumerate(co2_data): 
    year, CO2_level = data
    next_year, next_co2_level = co2_data[index+1]
    #print(next_year,next_co2_level)
    if index == 19: 
         break
    if next_co2_level - CO2_level >= 3: 
        print(f' {year} and {next_year}, {next_co2_level - CO2_level}')


 2015 and 2016, 3.410000000000025
